### PR TITLE
ROE-2272 Change uri, and validation to make email search agnostic oe …

### PR DESCRIPTION
**Pull request by mattch1**
…or not

### JIRA link
https://companieshouse.atlassian.net/browse/ROE-2272

### Change description
Allow a wider set of company number formats to be accepted
Return agnostic company email details object coupled with agnostic url.
Update validation pattern to match the generic one add to the private SDK for the company number in the url.

### Work checklist

- [x] Tests added where applicable

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java [hunk 1]
```diff
@@ -16,26 +16,26 @@
...
 public class PrivateDataRetrievalService {
 
-    private static final String OVERSEAS_ENTITY_URI_SECTION = "/overseas-entity/%s/entity-data";
+    private static final String OVERSEAS_ENTITY_URI_SECTION = "/company/%s/registered-email-address";
 
     @Autowired
     private ApiClientService apiClientService;
 
-    public OverseasEntityDataApi getOverseasEntityData(String requestId, String companyNumber)
+    public RegisteredEmailAddressJson getRegisteredEmailAddressData(String requestId, String companyNumber)
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:24]: Would name this simply `getRegisteredEmailAddress`.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java [hunk 2]
```diff
@@ -16,26 +16,26 @@
...
+            var registeredEmailAddress = apiClientService
                     .getInternalApiClient()
-                    .privateOverseasEntityDataHandler()
-                    .getOverseasEntityData(String.format(OVERSEAS_ENTITY_URI_SECTION, companyNumber))
+                    .privateCompanyResourceHandler()
+                    .getCompanyRegisteredEmailAddress(String.format(OVERSEAS_ENTITY_URI_SECTION, companyNumber))
                     .execute()
                     .getData();
 
             ApiLogger.infoContext(requestId, "Successfully retrieved overseas entity data from database", logDataMap.getLogMap());
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:37]: Log output can now be a bit more specific - "Successfully retrieved overseas entity email address from database".

File: src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java [hunk 3]
```diff
@@ -13,7 +13,7 @@ public class AuthCodeEmailValidator {
 
     public static final String AUTH_CODE_REGEX = "^[A-Z0-9]{6}$";
 
-    public static final String COMPANY_NUMBER_REGEX = "^[A-Za-z]{2}[0-9]{6}$";
+    public static final String COMPANY_NUMBER_REGEX = "^[^/]([A-Za-z0-9]{6,10})$";
```
Comment (mwestacott) [inline src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java:16]: Wondering why this has changed? Does it maybe match the reg exp in the new Oracle Query API end-point?

File: src/test/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidatorTest.java [hunk 4]
```diff
@@ -26,15 +25,10 @@ void setup() {
...
-        Errors errors = authCodeEmailValidator.validate("OE000001", "A1B2C3", new Errors(), CONTEXT);
-        assertFalse(errors.hasErrors());
-    }
-
-    @Test
-    void testSuccessfulValidationWhenCompanyNumberStartsWithTwoLowercaseLetters() {
-        Errors errors = authCodeEmailValidator.validate("oe001100", "A1B2C3", new Errors(), CONTEXT);
+    @ParameterizedTest
+    @ValueSource(strings = { "00110011", "ABCDEFGH","OE000001", "oe001100"})
+    void testSuccessfulValidationWhenCompanyNumberStartsWithTwoLowercaseLetters(String companyNumber) {
```
Comment (mwestacott) [inline src/test/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidatorTest.java:30]: Test method name doesn't match the test input strings. Should it just be named `testNoErrorsReportedWhenCompanyNumberIsValid`.

File: src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java (hunk 3 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java:16]: It matches the check I added to the SDK https://github.com/companieshouse/private-api-sdk-java/pull/311/files

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java [hunk 5]
```diff
@@ -16,26 +16,26 @@
...
 
-    public OverseasEntityDataApi getOverseasEntityData(String requestId, String companyNumber)
+    public RegisteredEmailAddressJson getOverseasEntityData(String requestId, String companyNumber)
             throws ServiceException {
         var logDataMap = new DataMap.Builder().companyNumber(companyNumber).build();
         try {
             ApiLogger.infoContext(requestId, "Retrieving overseas entity data from database", logDataMap.getLogMap());
 
-            var overseasEntityDataApi = apiClientService
+            var registeredEmailAddress = apiClientService
```
Comment (pjames-ch) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:30]: Is this use of var here and on line 26 to satisfy sonarqube? I really think we should not be losing type information - this is not a new change, but just clarifying. I will try to get the sonarqube rules updated if so.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java [hunk 6]
```diff
@@ -16,26 +16,26 @@
...
+                    .getCompanyRegisteredEmailAddress(String.format(OVERSEAS_ENTITY_URI_SECTION, companyNumber))
                     .execute()
                     .getData();
 
             ApiLogger.infoContext(requestId, "Successfully retrieved overseas entity data from database", logDataMap.getLogMap());
-            return overseasEntityDataApi;
+            return registeredEmailAddress;
         } catch (ApiErrorResponseException e) {
             if (e.getStatusCode() == 404) {
                 var message = "Unable to find overseas entity data from database, HTTP exception status: " + e.getStatusCode();
```
Comment (pjames-ch) [inline src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:41]: Here I guess the error case can also be more specific "overseas entity email address"

Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:41]: The ticket requires this to be more generic so keeping "overseas entity" out of it.

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 5 above)
Comment (pjames-ch) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java:30]: Ok thanks - I will chase the rule update as I feel this is just making code hard to follow in terms of typing information.

File: src/test/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidatorTest.java (hunk 4 above)
Comment (mattch1) [reply src/test/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidatorTest.java]: Changed name

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 2 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java]: Updated all messages to be more generic while at the same time specifying "company registered email address"

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 1 above)
Comment (mattch1) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java]: Renamed.

File: src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java (hunk 3 above)
Comment (mwestacott) [reply src/main/java/uk/gov/companieshouse/authcodenotification/validation/AuthCodeEmailValidator.java:16]: I think it's fine - just wasn't expecting it. Makes the service more generic so I'm fine with this.

**Review by mwestacott — APPROVED** [review]

File: src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java (hunk 6 above)
Comment (pjames-ch) [reply src/main/java/uk/gov/companieshouse/authcodenotification/service/PrivateDataRetrievalService.java]: Apologies - I was looking at this through the wrong lens.

**Review by pjames-ch — APPROVED** [review]
